specs/DHT.md

```diff
@@ -0,0 +1,399 @@
+
+## Protocol Specification
+
+### Data Types
+
+The data structures are defined in [pb/p2pd.proto](pb/p2pd.proto). All messages
+are varint-delimited. For the DHT queries, the relevant data types are:
+
+- `DHTRequest`
+- `DHTResponse`
+
+All DHT requests will be wrapped in a `Request` message with `Type: DHT`. All
+DHT responses from the daemon will be wrapped in a `Response` with the
+`DHTResponse` field populated.
+
+`DHTRequest` messages have a `Type` parameter that specifies the specific query
+the client wishes to execute.
+
+`DHTResponse` messages have a `Type` parameter that specifies whether a response
+marks the `BEGIN`ning of a stream of messages, a `VALUE` within a stream of
+messages, or the `END` of a stream of messages.
+
+### Protocol Requests
+
+*Protocols described in pseudo-go. Items of the form [item, ...] are lists of
+many items.*
+
+#### Errors
+
+Any response that may be an error, will take the form of:
+
+```
+Response{
+  Type: ERROR,
+  ErrorResponse: {
+    Msg: <error message>,
+  },
+}
+```
+
+#### `FIND_PEER`
+Clients can issue a `FIND_PEER` request to query the DHT for a given peer's
+known addresses.
+
+**Client**
+```
+Request{
+  Type: DHT,
+  DHTRequest: DHTRequest{
+    Type: FIND_PEER,
+    Peer: <peer id>,
+  },
+}
+```
+
+**Daemon**
+*Can return an error*
+
+```
+Response{
+  Type: OK,
+  DHTResponse: DHTResponse{
+    Type: Value,
+    Peer: PeerInfo{
+      Id: <peer id>,
+      Addrs: [<addr>, ...],
+    }
+  }
+}
+```
+
+#### `FIND_PEERS_CONNECTED_TO_PEER`
+Clients can issue a `FIND_PEERS_CONNECTED_TO_PEER` request to query the DHT for
+peers directly connected to a given peer in the DHT.
+
+**Client**
+```
+Request{
+  Type: DHT,
+  DHTRequest: DHTRequest{
+    Type: FIND_PEERS_CONNECTED_TO_PEER,
+    Peer: <peer id>,
+  },
+}
+```
+
+**Daemon**
+*Can return an error*
+
+```
+Response{
+  Type: OK,
+  DHTResponse: DHTResponse{
+    Type: BEGIN,
+  }
+}
+```
+
+**Daemon**
+*Can return any number of responses like this, including 0*
+
+```
+Response{
+  Type: OK,
+  DHTResponse: DHTResponse{
+    Type: VALUE,
+    Peer: PeerInfo{
+      Id: <peer id>,
+      Addrs: [<addr>, ...],
+    },
+  },
+}
+```
+
+**Daemon**
+*Marks the end of the result stream*
+
+```
+Response{
+  Type: OK,
+  DHTResponse: DHTResponse{
+    Type: END,
+  }
+}
+```
+
+#### `FIND_PROVIDERS`
+Clients can issue a `FIND_PROVIDERS` request to query the DHT for peers
+that have a piece of content, identified by a CID. `FIND_PROVIDERS` optionally
+include a `Count`, specifying a maximum number of results to return.
+
+**Client**
+```
+Request{
+  Type: DHT,
+  DHTRequest: DHTRequest{
+    Type: FIND_PROVIDERS,
+    Cid: <content id>,
+    Count: <number of results to include>,
+  },
+}
+```
+
+**Daemon**
+*Can return an error if the CID is invalid*
+
+```
+Response{
+  Type: OK,
+  DHTResponse: DHTResponse{
+    Type: BEGIN,
+  }
+}
+```
+
+**Daemon**
+*Can return any number of responses like this, including 0*
+
+```
+Response{
```

Comment:
the stream of values is not wrapped in `Response` objects, they are raw `DHTResponse`s.